src/r_data.c

```diff
@@ -966,150 +968,197 @@ int R_ColormapNumForName(const char *name)
-  else
-    {   // Compose a default transparent filter map based on PLAYPAL.
-      unsigned char *playpal = W_CacheLumpName("PLAYPAL", PU_STATIC);
-      char *fname = M_StringJoin(D_DoomPrefDir(), DIR_SEPARATOR_S, "tranmap.dat");
-      struct {
-        unsigned char pct;
-        unsigned char playpal[256*3]; // [FG] a palette has 256 colors saved as byte triples
-      } cache;
-      FILE *cachefp = M_fopen(fname,"r+b");
-
-      if (main_tranmap == NULL) // [FG] prevent memory leak
-      {
-      main_tranmap = Z_Malloc(256*256, PU_STATIC, 0);  // killough 4/11/98
-      }
+  for (int i = 0; i < sizeof(playpal_digest); ++i)
+  {
+    sprintf(&playpal_string[i * 2], "%02x", playpal_digest[i]);
+  }
+  playpal_string[32] = '\0';
+}
 
-      // Use cached translucency filter if it's available
+static void CreateTranMapBaseDir(void)
+{
+  const char* data_root = D_DoomPrefDir();
+  const int32_t length = strlen(data_root) + 10; // "/tranmaps\0"
 
-      if (!cachefp ? cachefp = M_fopen(fname,"w+b") , 1 : // [FG] open for writing and reading
-          fread(&cache, 1, sizeof cache, cachefp) != sizeof cache ||
-          cache.pct != tran_filter_pct ||
-          memcmp(cache.playpal, playpal, sizeof cache.playpal) ||
-          fread(main_tranmap, 256, 256, cachefp) != 256 ||  // killough 4/11/98
-          force_rebuild)
-        {
-          long pal[3][256], tot[256], pal_w1[3][256];
-          long w1 = ((unsigned long) tran_filter_pct<<TSC)/100;
-          long w2 = (1l<<TSC)-w1;
+  tranmap_dir = Z_Malloc(length, PU_STATIC, 0);
+  snprintf(tranmap_dir, length, "%s/tranmaps", data_root);
 
-          // First, convert playpal into long int type, and transpose array,
-          // for fast inner-loop calculations. Precompute tot array.
+  M_MakeDirectory(tranmap_dir);
+}
 
-          {
-            register int i = 255;
-            register const unsigned char *p = playpal+255*3;
-            do
-              {
-                register long t,d;
-                pal_w1[0][i] = (pal[0][i] = t = p[0]) * w1;
-                d = t*t;
-                pal_w1[1][i] = (pal[1][i] = t = p[1]) * w1;
-                d += t*t;
-                pal_w1[2][i] = (pal[2][i] = t = p[2]) * w1;
-                d += t*t;
-                p -= 3;
-                tot[i] = d << (TSC-1);
-              }
-            while (--i>=0);
-          }
+static void CreateTranMapPaletteDir(void)
+{
+  if (!tranmap_dir)
+    CreateTranMapBaseDir();
 
-          // Next, compute all entries using minimum arithmetic.
+  if (!playpal_string[0])
+    CalculatePlaypalChecksum();
 
-          {
-            int i,j;
-            byte *tp = main_tranmap;
-            for (i=0;i<256;i++)
-              {
-                long r1 = pal[0][i] * w2;
-                long g1 = pal[1][i] * w2;
-                long b1 = pal[2][i] * w2;
+  // Using vanilla's playpal checksum as example
+  int32_t length = strlen(tranmap_dir) + sizeof("/4804c7f34b5285c334a7913dd98fae16");
```

Comment:
That's `sizeof(playpal_string)`. The array does not need to be filled to have its size defined (is not `strlen() + 1`).